### PR TITLE
Track Job completion through Pod finalizers and status

### LINKED ISSUE #28486: Job controller should not rely on calculating pod for its status
Currently [our job controller](https://github.com/kubernetes/kubernetes/blob/35a345269eac879b0d9f25110d598845dcf8a62f/pkg/controller/job/controller.go#L452-L456) relies on completed and failed pods to stay around to be properly calculated. This might become cumbersome when a job will be big (iow. high completions number). It would be nice to re-implement this part so that those status numbers are updated on each run rather than re-calculated each time.

@erictune you've mentioned this some time ago, it would be nice to fix it some time, that's why I'll create this issue so we don't forget.

### LINKED ISSUE #99368: Track completed indexes without relying on lingering pods
The features "Indexed Job" kubernetes/enhancements#2214 and "Job tracking without lingering pods" kubernetes/enhancements#2307 won't work together in alpha. But it is a requirement for beta graduation.

/sig apps
/area workload-api/job
/assign

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Tracks Job completion without depending on Pods to not be removed. Instead, it keeps a finalizer on the Pods until they have been accounted into the status.

**Which issue(s) this PR fixes**:

Fixes #28486
Fixes #99368

**Special notes for your reviewer**:

Builds on top of #98817 for API changes.
Implements kubernetes/enhancements#2307 with the update from kubernetes/enhancements#2585

**Does this PR introduce a user-facing change?**:

```release-note
Track Job completion through status and Pod finalizers, removing dependency on Pod tombstones.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: http://git.k8s.io/enhancements/keps/sig-apps/2307-job-tracking-wihout-lingering-pods
```

### CODE REVIEW

File: pkg/controller/job/indexed_job_utils.go [hunk 1]
```diff
@@ -38,82 +38,173 @@ func isIndexedJob(job *batch.Job) bool {
...
+	Last  int
+}
+
+type orderedIntervals []interval
+
+// calculateSucceededIndexes returns the list of succeeded indexes in
+// in compressed format (intervals).
+// The first list only contains the indexes in .status.completedIndexes if
+// tracking with finalizers. The second list includes the indexes that
+// succeeded since the last sync.
```
Comment (soltysh) [inline pkg/controller/job/indexed_job_utils.go:52]: Maybe just say it returns before and after intervals for ease of creating a patch?

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -646,32 +724,240 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+// 3. Increment job counters for pods that no longer have a finalizer.
+// 4. Add Complete condition if satisfied with current counters.
+// It does this in a controlled way such that the size of .status doesn't grow
+// too much.
+func (jm *Controller) trackJobStatusAndRemoveFinalizers(job *batch.Job, pods []*v1.Pod, succeededIndexes orderedIntervals, uncounted uncountedTerminatedPods, finishedCond *batch.JobCondition, needsFlush bool) error {
+	isIndexed := isIndexedJob(job)
+	var podsToRemoveFinalizer []*v1.Pod
+	uncountedStatus := job.Status.UncountedTerminatedPods
+	var newSucceededIndexes []int
+	flushUncountedAndRemoveFinalizers := func(last bool) error {
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:815]: Any particular reason for embedding this rather big function? I prefer to use function embedding only for small ones, it's easier to read w/o embedding, imo.

Comment (soltysh) [reply pkg/controller/job/job_controller.go:815]: Yeah, reading through the code it's unnecessarily complicated. I'd prefer this being separate method with clear inputs and outputs, the current approach makes it hard to follow the logic.

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -646,32 +724,240 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+		if len(newUncounted) != len(uncountedStatus.Failed) {
+			needsFlush = true
+			job.Status.Failed += int32(len(uncountedStatus.Failed) - len(newUncounted))
+			uncountedStatus.Failed = newUncounted
+		}
+		if rmErr == nil && last && jm.enactJobFinished(job, finishedCond) {
+			needsFlush = true
+		}
+		if needsFlush && (last || rmErr != nil) {
+			if err := jm.updateStatusHandler(job); err != nil {
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:852]: As mentioned before, there should be clear inputs, reaction to it (this method) and status update and return. Currently there are possibly 2 status updates, which might be unnecessary.

**Review by soltysh — CHANGES_REQUESTED** [review]
I did a first pass, I'll do another after fixing the initial comments.

File: pkg/controller/job/job_controller.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go]: See the update. But to be clear, both status updates are necessary.

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -646,32 +724,276 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+		}
+		job.Status.CompletionTime = &finishedCond.LastTransitionTime
+		jm.recorder.Event(job, v1.EventTypeNormal, "Completed", "Job completed")
+	} else {
+		jm.recorder.Event(job, v1.EventTypeWarning, finishedCond.Reason, finishedCond.Message)
+	}
+	return true
+}
+
+func updateUncountedWithFailedFinalizerRemovals(uncounted []types.UID, uidsWithFinalizer sets.String) []types.UID {
```
Comment (ahg-g) [inline pkg/controller/job/job_controller.go:989]: nit: rename the function, we are not updating anything

**Review by ahg-g — COMMENTED** [review]
First round, I would like to make another one. 

A couple of notes:
- It seems that a lot of the changes were introduced to special case indexed job, I wonder if the PR would have been simpler if we did that as a followup (not asking to split now)?
- the controller can use a refactoring, but perhaps it should be done after the various features graduate to GA.

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -646,32 +724,276 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+				needsFlush = true
+				uncountedStatus.Succeeded = append(uncountedStatus.Succeeded, pod.UID)
+			}
+		} else if pod.Status.Phase == v1.PodFailed || podTerminating {
+			if !uncounted.failed.Has(string(pod.UID)) && (!isIndexed || getCompletionIndex(pod.Annotations) != unknownCompletionIndex) {
+				needsFlush = true
+				uncountedStatus.Failed = append(uncountedStatus.Failed, pod.UID)
+			}
+		}
+		if len(uncountedStatus.Succeeded) >= maxUncountedPods || len(uncountedStatus.Failed) >= maxUncountedPods {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:847]: It's a heuristic anyways. But yeah, your suggestion makes more sense.

File: pkg/controller/job/job_controller.go (hunk 4 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:989]: probably another leftover from a previous approach 😅

**Review by alculquicondor — COMMENTED** [review]
> It seems that a lot of the changes were introduced to special case indexed job, I wonder if the PR would have been simpler if we did that as a followup

I had them split initially, but the indexed job implementation changed a lot of the same lines. So I ended up merging them.

> the controller can use a refactoring, but perhaps it should be done after the various features graduate to GA.

I agree. I would make it part of the graduation, even.

File: pkg/controller/job/job_controller.go [hunk 6]
```diff
@@ -622,16 +663,52 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
 
+	if uncounted != nil {
+		needsStatusUpdate := suspendCondChanged || active != job.Status.Active
+		job.Status.Active = active
+		err = jm.trackJobStatusAndRemoveFinalizers(&job, pods, prevSucceededIndexes, *uncounted, finishedCondition, needsStatusUpdate)
+		if err != nil {
+			return false, err
+		}
+		jobFinished := IsJobFinished(&job)
+		if jobHaveNewFailure && !jobFinished {
```
Comment (ahg-g) [inline pkg/controller/job/job_controller.go:674]: `jobHasNewFailure`?

File: pkg/controller/job/indexed_job_utils.go [hunk 7]
```diff
@@ -38,85 +38,181 @@ func isIndexedJob(job *batch.Job) bool {
...
+	withFinalizers := trackingUncountedPods(job)
+	if withFinalizers {
+		prevIntervals = succeededIndexesFromJob(job)
+	}
+	newSucceeded := sets.NewInt()
 	for _, p := range pods {
 		ix := getCompletionIndex(p.Annotations)
-		if ix == unknownCompletionIndex {
-			continue
+		if ix != unknownCompletionIndex && ix < int(*job.Spec.Completions) && p.Status.Phase == v1.PodSucceeded && (!withFinalizers || hasJobTrackingFinalizer(p)) {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/indexed_job_utils.go]: Done.

File: pkg/controller/job/job_controller.go [hunk 8]
```diff
@@ -646,32 +724,276 @@ func (jm *Controller) syncJob(key string) (bool, error) {
...
+// Returns number of successfully deleted pods.
+func (jm *Controller) deleteActivePods(job *batch.Job, pods []*v1.Pod) (int32, error) {
+	errCh := make(chan error, len(pods))
+	successfulDeletes := int32(len(pods))
+	wg := sync.WaitGroup{}
+	wg.Add(len(pods))
+	for i := range pods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+			if err := jm.podControl.DeletePod(job.Namespace, pod.Name, job); err != nil && !apierrors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/controller/job/job_controller.go:757]: it will set the deletion timestamp, right?

**Review by ahg-g — COMMENTED** [review]
LGTM 

module the two open comments.

File: pkg/controller/job/job_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:757]: Yes, updated the method's comment.

File: pkg/controller/job/job_controller.go [hunk 9]
```diff
@@ -725,23 +1047,31 @@ func newCondition(conditionType batch.JobConditionType, status v1.ConditionStatu
...
+func getStatus(job *batch.Job, pods []*v1.Pod, uncounted *uncountedTerminatedPods) (succeeded, failed int32) {
+	if uncounted != nil {
+		succeeded = job.Status.Succeeded + int32(uncounted.Succeeded().Len())
+		failed = job.Status.Failed + int32(uncounted.Failed().Len())
+	}
+	succeeded += int32(countValidPodsWithFilter(job, pods, uncounted.Succeeded(), func(p *v1.Pod) bool {
+		return p.Status.Phase == v1.PodSucceeded
+	}))
+	failed += int32(countValidPodsWithFilter(job, pods, uncounted.Failed(), func(p *v1.Pod) bool {
+		return p.Status.Phase == v1.PodFailed || (p.DeletionTimestamp != nil && p.Status.Phase != v1.PodSucceeded)
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:1092]: ping @smarterclayton

**Review by soltysh — COMMENTED** [review]
I'll do another pass after API gets approved

File: pkg/controller/job/job_controller.go [hunk 10]
```diff
@@ -433,8 +447,31 @@ func (jm *Controller) getPodsForJob(j *batch.Job) ([]*v1.Pod, error) {
...
+	for i, p := range pods {
+		adopted := true
+		for _, r := range p.OwnerReferences {
+			if r.UID == j.UID {
+				adopted = false
+				break
+			}
+		}
+		if adopted && !hasJobTrackingFinalizer(p) {
+			pods[i] = p.DeepCopy()
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:474]: sgtm

File: pkg/controller/job/job_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go:1094]: Ack

File: pkg/controller/job/indexed_job_utils_test.go [hunk 11]
```diff
@@ -98,30 +106,175 @@ func TestCalculateSucceededIndexes(t *testing.T) {
...
+			},
+			trackingWithFinalizers: true,
+			wantStatusIntervals: []interval{
+				{2, 7},
+			},
+			wantIntervals: []interval{
+				{0, 0},
+				{2, 7},
+				{9, 9},
+			},
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/indexed_job_utils_test.go:201]: Done

**Comment by soltysh** [comment]
This still looks reasonably, please ping me when the API gets merged and this will be ready for final review.

File: pkg/controller/job/job_controller.go [hunk 12]
```diff
@@ -972,18 +1324,112 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
...
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": []string{batch.JobTrackingFinalizer},
+		},
+	}
+	patchBytes, _ := json.Marshal(patch)
+	return patchBytes
+}
+
+func removeTrackingFinalizerPatch(obj metav1.Object) []byte {
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go]: Nit: unless I missed but I see this being executed only against pods, so you could change the signature to just pods. Same for `hasJobTrackingFinalizer`.

File: pkg/controller/job/job_controller.go [hunk 13]
```diff
@@ -127,6 +130,7 @@ func NewController(podInformer coreinformers.PodInformer, jobInformer batchinfor
 		},
 		expectations: controller.NewControllerExpectations(),
 		queue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(DefaultJobBackOff, MaxJobBackOff), "job"),
+		orphanQueue:  workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(DefaultJobBackOff, MaxJobBackOff), "joborphanpod"),
```
Comment (soltysh) [inline pkg/controller/job/job_controller.go:133]: Nit: `job_orphan_pods` for the queue name.

File: pkg/controller/job/indexed_job_utils.go [hunk 14]
```diff
@@ -39,85 +39,183 @@ func isIndexedJob(job *batch.Job) bool {
...
+		j++
 	}
-	return result.String(), count
+	return result
 }
 
-func addSingleOrRangeStr(builder *strings.Builder, first, last int) {
-	if builder.Len() > 0 {
-		builder.WriteRune(',')
+func (oi orderedIntervals) total() int {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/job/indexed_job_utils.go:110]: Done

File: pkg/controller/job/job_controller.go (hunk 12 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go]: It was for jobs and pods at some point. Reverted.

File: pkg/controller/job/job_controller.go (hunk 13 above)
Comment (alculquicondor) [reply pkg/controller/job/job_controller.go]: Done

**Review by alculquicondor — COMMENTED** [review]
Squashed (minus the API review commits)

**Review by soltysh — APPROVED** [review]
/triage accepted
/priority backlog
/lgtm
/approve

**Comment by alculquicondor** [comment]
Rebased on top of reviewed API changes

File: pkg/controller/controller_ref_manager.go [hunk 15]
```diff
@@ -565,6 +572,7 @@ func ownerRefControllerPatch(controller metav1.Object, controllerKind schema.Gro
 					BlockOwnerDeletion: &blockOwnerDeletion,
 				},
 			},
+			Finalizers: finalizers,
```
Comment (liggitt) [inline pkg/controller/controller_ref_manager.go:575]: does this add or overwrite the whole list?

Comment (alculquicondor) [reply pkg/controller/controller_ref_manager.go:575]: adds, due to the patchStrategy

File: pkg/controller/job/indexed_job_utils.go [hunk 16]
```diff
@@ -39,85 +39,184 @@ func isIndexedJob(job *batch.Job) bool {
...
+}
+
+type orderedIntervals []interval
+
+// calculateSucceededIndexes returns the old and new list of succeeded indexes
+// in compressed format (intervals).
+// The old list is solely based off .status.completedIndexes, but returns an
+// empty list if this Job is not tracked with finalizers. The new list includes
+// the indexes that succeeded since the last sync.
+func calculateSucceededIndexes(job *batch.Job, pods []*v1.Pod) (orderedIntervals, orderedIntervals) {
```
Comment (liggitt) [inline pkg/controller/job/indexed_job_utils.go:54]: is there a way to isolate this change so the existing logic is used if the feature is disabled? it's hard to reason about the safety of this change for existing jobs when the new feature is off

Comment (alculquicondor) [reply pkg/controller/job/indexed_job_utils.go:54]: the way would be to revert all the changes and add alternate implementations. I could do that, but we will probably miss the code freeze deadline. I added a lot of coverage for this.

**Comment by alculquicondor** [comment]
Rebased

**Comment by liggitt** [comment]
/approve

adding patch to a controller that already has update lgtm

**Comment by JamesLaverack** [comment]
@alculquicondor @ahg-g Why is the hold here? This PR is approved otherwise (and would therefore make https://github.com/kubernetes/enhancements/issues/2307 good for code freeze).

**Comment by alculquicondor** [comment]
Oops, I forgot to remove the hold. It was on hold for the API PR, which is now merged.
